kernel/fatal.c

```diff
@@ -0,0 +1,92 @@
+#include <kernel.h>
+#include <kernel_internal.h>
+#include <misc/printk.h>
+#include <arch/cpu.h>
+#include <logging/log_ctrl.h>
+#include <fatal.h>
+
+/* LCOV_EXCL_START */
+FUNC_NORETURN __weak void z_arch_system_halt(unsigned int reason)
```

Comment:
I'd put the arch_irq_lock() in here, not outside the call.   Masked interrupts are sort of part of the definition of a system halt, someone who calls this expecting a halt but forgets the lock is going to be surprised when their timeouts & driver I/O keep happening.

Also, needless to say this is going to work only with uniprocessor setups. On SMP systems, halting the system is more complicated.  We could do it portably on top of the existing sched_ipi() mechanism (basically calling this function out of the IPI handler, which itself already has a fallback for systems without a working IPI), which would probably be useful here if we want a robust weak fallback.

Alternatively, you could #ifndef CONFIG_SMP around the weak fallback and force the SMP systems to do this themselves.  But the scheme here leaves them with a silently broken halt.

Reply:
moved.